Always run in the directory of the file when no workspace

mjcrouch opened the pull request:
This slightly improves the `directoryOverride` by centralizing the logic, without having it spread over multiple places.

There were only a few usages of directoryOverride and they were all doing the same thing - if and only if there were no workspace folders, it was adding the -d option to commands, using the directory of the relevant file (this is a very difficult situation to get into, as the extension does not activate if there is no folder opened)

Perforce uses -d to override the cwd / PWD environment - so instead we can just spawn the process in the right place with the right environment. Tested working on linux; I still need to test this on windows.

Now, it always uses the directory if there is no workspace for the file, i.e. if you open a file that is not in your workspace, you should still be able to perform perforce commands on it, if a default p4 command would work in that directory.

The existing behaviour for working out where to run commands is not ideal - it looks for a configuration assigned to a workspace, meaning that multiple p4 configurations in a single workspace is not necessarily working - see issue #41

This does not attempt to resolve that, just deals with removing the directoryOverride option

allender commented:
I've been meaing to ask about this.  Why is there directory override.  If p4 commands are run on full pathnames, does the -d option even need to get used?

mjcrouch commented:
well, I removed it - I decided to have a look into it after I noticed the diff on your PR.
Previously for files outside of the workspace it was actually running with a cwd of `""`, instead of the file location. The file location makes more sense because perforce has a better chance of working out what to do with it.

And as far as I can tell from the docs / experimenting, -d is effectively the same as running in that directory, so I just changed it to run in that location instead.

mjcrouch commented on src/PerforceService.ts:
```diff
@@ -240,7 +232,13 @@ export namespace PerforceService {
         }
 
-        const spawnArgs: CP.SpawnOptions = { cwd: config ? config.localDir : wksPath };
+        const cwd = config
```
Note to self- use `??` Or `||` here instead

allender replied:
In 13 years of using perforce I never needs the -d global option on the command line. Original code seems to indicate it was because of multi roots in the vscode workspace but I don’t understand that either. Are workspaces and the scm client related somehow?

mjcrouch replied:
Because perforce may infer the client, user, port settings etc from the directory where you run the command, if the directory or one of its parents contains a p4config file. 

So if you open a file in vscode from outside of your vscode workspace, and that file is under a dir with a p4config, by running in the file's directory we can still use perform actions on the file
